pkg/state/appender.go

```diff
@@ -630,6 +629,126 @@ func (a *txAppender) createInitialBlockSnapshot(minerAndRewardDiff txDiff) (txSn
+	}
+	return nil
+}
+
+func (a *txAppender) appendTxs(
+	params *appendBlockParams,
+	info *checkerInfo,
+	stateHash crypto.Digest,
+	hasher *txSnapshotHasher,
+) error {
+	blockInfo, errBlockInfo := a.currentBlockInfo()
+	if errBlockInfo != nil {
+		return errBlockInfo
+	}
+	blockV5Activated, err := a.stor.features.newestIsActivated(int16(settings.BlockV5))
+	if err != nil {
+		return err
+	}
+	consensusImprovementsActivated, err := a.stor.features.newestIsActivated(int16(settings.ConsensusImprovements))
+	if err != nil {
+		return err
+	}
+	blockRewardDistributionActivated, err := a.stor.features.newestIsActivated(int16(settings.BlockRewardDistribution))
+	if err != nil {
+		return err
+	}
+	invokeExpressionActivated, err := a.stor.features.newestIsActivated(int16(settings.InvokeExpression))
+	if err != nil {
+		return err
+	}
+	// Check and append transactions.
+	var bs proto.BlockSnapshot
+
+	for _, tx := range params.transactions {
+		appendTxArgs := &appendTxParams{
+			chans:                            params.chans,
+			checkerInfo:                      info,
+			blockInfo:                        blockInfo,
+			block:                            params.block,
+			acceptFailed:                     blockV5Activated,
+			blockV5Activated:                 blockV5Activated,
+			rideV5Activated:                  info.rideV5Activated,
+			rideV6Activated:                  info.rideV6Activated,
+			consensusImprovementsActivated:   consensusImprovementsActivated,
+			blockRewardDistributionActivated: blockRewardDistributionActivated,
+			invokeExpressionActivated:        invokeExpressionActivated,
+			validatingUtx:                    false,
+			currentMinerPK:                   params.block.GeneratorPublicKey,
+		}
+		txSnapshots, errAppendTx := a.appendTx(tx, appendTxArgs)
+		if errAppendTx != nil {
+			return errAppendTx
+		}
+		bs.AppendTxSnapshot(txSnapshots.regular)
+
+		txID, idErr := tx.GetID(a.settings.AddressSchemeCharacter)
+		if idErr != nil {
+			return idErr
+		}
+
+		if len(txSnapshots.regular) == 0 { // sanity check
+			return errors.Errorf("snapshot of txID %q cannot be empty", base58.Encode(txID))
+		}
+		txSh, shErr := calculateTxSnapshotStateHash(hasher, txID, blockInfo.Height, stateHash, txSnapshots.regular)
+		if shErr != nil {
+			return errors.Wrapf(shErr, "failed to calculate tx snapshot hash for txID %q at height %d",
+				base58.Encode(txID), blockInfo.Height,
+			)
+		}
+		stateHash = txSh // update stateHash in order to accumulate state hashes into block snapshot hash
+	}
+	blockID := params.block.BlockID()
+	if ssErr := a.stor.snapshots.saveSnapshots(blockID, blockInfo.Height, bs); ssErr != nil {
+		return ssErr
+	}
+	// TODO: check snapshot hash with the block snapshot hash if it exists
+	if shErr := a.stor.stateHashes.saveSnapshotStateHash(stateHash, blockInfo.Height, blockID); shErr != nil {
+		return errors.Wrapf(shErr, "failed to save block shasnpt hash at height %d", blockInfo.Height)
+	}
+	return nil
```

Comment:
This code is common for `appendTxs` and `applySnapshotsInLightNode` methods. I think it should be placed in `txAppender.appendBlock` method.